salt/utils/network.py

```diff
@@ -71,207 +72,78 @@ def host_to_ip(host):
-        - FQDN
-        - public ipaddress
-        - localhost alias
-        - private ipaddress
-    '''
-    # punish matches in order of preference
-    punish = [
-        'localhost.localdomain',
-        'localhost.my.domain',
-        'localhost4.localdomain4',
-        'localhost',
-        'ip6-localhost',
-        'ip6-loopback',
-        'ipv6-localhost',
-        'ipv6-loopback',
-        '127.0.2.1',
-        '127.0.1.1',
-        '127.0.0.1',
-        '0.0.0.0',
-        '::1',
-        'fe00::',
-        'fe02::',
-    ]
-
-    def _key_hostname(e):
-        # should never have a space in hostname
-        # favor hostnames w/o spaces
-        if ' ' in e:
-            first = 1
-        else:
-            first = -1
-
-        # punish localhost list
-        if e in punish:
-            second = punish.index(e)
-        else:
-            second = -1
-
-        # punish ipv6
-        third = e.count(':')
-
-        # punish ipv4
-        # punish ipv4 addresses that start with '127.' more
-        e_is_ipv4 = e.count('.') == 3 and not any(c.isalpha() for c in e)
-        if e_is_ipv4:
-            if e.startswith('127.'):
-                fourth = 1
-            else:
-                fourth = 0
-        else:
-            fourth = -1
-
-        # favor hosts with more dots
-        fifth = -(e.count('.'))
-
-        # favor longest fqdn
-        sixth = -(len(e))
-
-        return (first, second, third, fourth, fifth, sixth)
-
-    return sorted(hostname_list, key=_key_hostname)
-
-
-def get_hostnames():
-    '''
-    Get list of hostnames using multiple strategies
+def _generate_minion_id():
     '''
-    h = []
-    h.append(socket.gethostname())
-    h.append(socket.getfqdn())
+    Get list of possible host names and convention names.
 
-    # try socket.getaddrinfo
-    try:
-        addrinfo = socket.getaddrinfo(
-            socket.gethostname(), 0, socket.AF_UNSPEC, socket.SOCK_STREAM,
-            socket.SOL_TCP, socket.AI_CANONNAME
-        )
-        for info in addrinfo:
-            # info struct [family, socktype, proto, canonname, sockaddr]
-            if len(info) >= 4:
-                h.append(info[3])
-    except socket.gaierror:
-        pass
-
-    # try /etc/hostname
-    try:
-        name = ''
-        with salt.utils.fopen('/etc/hostname') as hfl:
-            name = hfl.read()
-        h.append(name)
-    except (IOError, OSError):
-        pass
+    :return:
+    '''
+    # There are three types of hostnames:
+    # 1. Network names. How host is accessed from the network.
+    # 2. Host aliases. They might be not available in all the network or only locally (/etc/hosts)
+    # 3. Convention names, an internal nodename.
 
-    # try /etc/nodename (SunOS only)
-    if salt.utils.is_sunos():
+    class DistinctList(list):
+        '''
+        List, which allows to append only distinct objects.
+        Needs to work on Python 2.6, because of collections.OrderedDict only since 2.7 version.
```

Comment:
Thanks, I know actually. But the main idea is to have a helper list object that on the fly is 

- filtering items
- contains only _list items_
- and is fast enough in order to avoid unrelated if/else for filtering in multiple places or iterate over the things all over again few times (as in an old implementation).

Hence this is why I put this as an internal hidden class. With the `salt.utils.odict` the implementation I still have to add missing parts while then it gets slightly worse, because it requires also value (`None` in my case) and it has unnatural `set_keys` as I am working with only a list. As an example:

```python
import re
from salt.utils.odict import OrderedDict as odict

class Hosts(odict):
    localhost_matchers = ['localhost.*', 'ip6-.*', '127.*', r'0\.0\.0\.0', '::1.*', 
                          'ipv6-.*', 'fe00::.*', 'fe02::.*', '1.0.0.*.ip6.arpa']

    def __setitem__(self, key, value, dict_setitem=dict.__setitem__):
        for rgx in self.localhost_matchers:
            if re.match(rgx, key):
                return
        super(Hosts, self).__setitem__(key, value, dict_setitem)

    def set(self, key, value=None):
        self[key] = value
        return self

    def set_keys(self, iterable):
        for obj in iterable:
            self.set(obj)
        return self

    def first(self):
        return self.keys() and self.keys()[0] or None

# Example usage
h = Hosts()
h.set('localhost').set('foo').set('bar').set_keys(['foo', 'bar', 'spam', 'toto'])
print h
print h.first()
```